malleefowl/custom_workflow.py

```diff
@@ -0,0 +1,309 @@
+                }
+            }
+        },
+        "group_of_task_schema" : {
+            "type": "object",
+            "description": "Describe a group of tasks to be run concurrently",
+            "required": ["name", "max_processes", "map", "reduce", "tasks"],
+            "additionalProperties": False,
+            "properties": {
+                "name": {
+                    "description": "Group of task name",
+                    "type": "string"
+                },
+                "max_processes": {
+                    "description": "Number of processes to run concurrently to process the data",
+                    "type": "number",
+                    "minimum": 1
+                },
+                "map": {
+                    "oneOf": [
+                        { "$ref": "#/definitions/input_description_schema" },
+                        {
+                            "description": "Array of data that has to be mapped directly",
+                            "type": "array",
+                            "minItems": 1,
+                            "items": {
+                                "type": "string"
+                            }
+                        }
+                    ]
+                },
+                "reduce": { "$ref": "#/definitions/input_description_schema" },
+                "tasks": {
+                    "description": "Array of workflow task to run concurrently inside the group",
+                    "type": "array",
+                    "minItems": 1,
+                    "items": { "$ref": "#/definitions/workflow_task_schema" }
+                }
+            }
+        },
+        "input_description_schema" : {
+            "description": "Description of an input source",
+            "type": "object",
+            "required": ["task"],
+            "additionalProperties": False,
+            "properties": {
+                "task": {
+                    "description": "Task name",
+                    "type": "string"
+                },
+                "output": {
+                    "description": "Task output name",
+                    "type": "string"
+                },
+                "as_reference": {
+                    "description": "Specify if the task output should be obtained as a reference or not",
+                    "type": "boolean"
+                }
+            }
+        }
+    }
+}
+
+
+def run(workflow, monitor=None, headers=None):
+    """
+    Run the given workflow
+    :param workflow: json structure describing the workflow
+    :param monitor: monitor callback to receive messages and progress
+    :param headers: Headers to use when making a request to the WPS
+    :return: A summary of the execution which is a list of all task's xml status
+    """
+
+    try:
+        jsonschema.validate(workflow, workflow_schema)
+    except jsonschema.ValidationError as e:
+        raise Exception('The workflow is invalid : {0}'.format(str(e)))
```

Comment:
Any uncaught exception raises until it is catched by pywps (where the process handler is called). At this point PyWPS update the xml status with a Failure and the failure message is the one contained by the exception. This is why I didn't care with Exception type since only the message is carried over. But I agree that for code maintenance a WorkflowException is better